### ISSUE TITLE
Nested describes don't run if parent describe has .only (mocha 3)

### ISSUE THREAD

**Issue by jhartikainen** [mochajs/mocha#2378]
Previously, if you had a nested describe and tests inside a parent describe with .only, the child-describe blocks would also run.

It seems https://github.com/mochajs/mocha/issues/1481 may have broken this behavior. While I think the actual fix in 1481 makes sense, this side-effect of it seems counter-intuitive, and at least goes against my expectations of how this should work. It's also different from how this works in some other testing tools I've used in the past with similar features.

## Steps to reproduce

1. Create a test file with a root describe using .only.
2. Put another describe inside it, with an it inside
3. Run test

## Expected result

The tests inside the nested describe run

## Actual result

Only tests immediately inside the parent describe block run, at least with mocha@3.0.0-1

## Sample code

```javascript
describe.only('this should run', function() {
  it('an it should also run', function() { });

  describe('this nested describe should run', function() {
    it('and this nested it should run as well', function() { });
  });
});
```

Output:
```
  this should run
    ✓ an it should also run


  1 passing (14ms)
```

Expected output:
```

  this should run
    ✓ an it should also run
    this nested describe should run
      ✓ and this nested it should run as well


  2 passing (17ms)
```

**Comment by boneskull** [comment]
Yeah, OK, this was clearly the intent of #1481.  See [this line](https://github.com/mochajs/mocha/blob/8a754344db3d4d85780572d38ada272b1faa6ee6/test/acceptance/misc/only/bdd.js#L107-L107).

I'll need to re-read the discussion around it, because I wouldn't expect this behavior.

**Comment by boneskull** [comment]
Hmm.  Actually I'm not sure this particular case got coverage.

**Comment by boneskull** [comment]
This is a bit tricky, so I'm not going to finish it tonight; could use some assistance.  Basically what needs to happen is if a suite is marked "only", then *unless a test within that suite is marked "only"*, all child suites' tests should run as well.  Or, that's what I'd expect, anyhow.

```js
describe.only('foo', function () {
  describe('bar', function () {
    it('should run', function () {});
  });
});
```

```js
// do we agree?
describe.only('foo', function () {
  it.only('should run', function () {});
  describe('bar', function () {
    it('should not run', function () {});
  });
});
```

```js
// careful with the order...
describe.only('foo', function () {
  describe('bar', function () {
    it('should not run', function () {});
  });
  it.only('should run', function () {});
});
```

also deeply-nested suites may complicate matters, as you can't simply check the most immediate parent to see if it has tests marked "only";  when child suites are created, they should be given a `false` `isOnly` prop (regardless of whether `describe.only()` was used!) when the parent has a truthy `onlyTests` prop OR when the *parent* has a `false` `isOnly` prop.   Given that the default value of `isOnly` is `undefined`, one could think of this as a ternary flag...

**Comment by boneskull** [comment]
cc @mochajs/core 

this will delay the final v3 release until it's resolved or we can agree that this is what should be happening, which is unlikely...

**Comment by boneskull** [comment]
@jhartikainen If you're feeling generous, I'd encourage you to pull down the `issue/2378` branch and see if the behavior matches what you expect.

**Comment by jhartikainen** [comment]
@boneskull yep appears to function as expected on that branch

**Comment by boneskull** [comment]
closed by 9ea2011, but I'm not 100% convinced, so we'll see if this stays closed.
